utils/storage.py

```diff
@@ -0,0 +1,229 @@
+        for root, dirs, files in os.walk(data.path):
+            root_uri = f"file://{root}/" if root[-1] != '/' else f"file://{root}"
+            for file in files:
+                file_objects.append(f"{root_uri}{file}")
+
+    if not recursive:
+        for file in os.listdir(data.path):
+            if os.path.isfile(os.path.join(data.path, file)):
+                root_uri = f"file://{data.path}/" if data.path[-1] != '/' else f"file://{data.path}"
+                file_objects.append(f"{root_uri}{file}")
+
+    return file_objects
+
+
+def get_page_content(data_uri: str, **kwargs):
+    data = parse_data_uri(data_uri)
+    if data.store == 's3':
+        return get_s3_contents(data)
+    elif data.store == 'file':
+        return get_file_contents(data)
+    else:
+        raise Exception(f"Unknown data store: {data.store}")
+
+
+def get_s3_contents(data: DataStorage, **kwargs):
+    """
+    Get the body of the object located at data.path
+    """
+    s3 = boto3.client('s3', **kwargs)
+
+    try:
+        obj = s3.get_object(Bucket=data.bucket, Key=data.path)
+        return obj['Body'].read().decode('utf-8')
+    except Exception as e:
+        url = (
+            f"https://{data.bucket}.s3.us-west-2.amazonaws.com/"
+            f"index.html#{data.path}/"
+        )
+        raise Exception(
+            f"Error reading file at {data.uri}\nCheck: {url}\n{e}"
+        )
+
+
+def get_file_contents(data: DataStorage):
+    """
+    Get the body of the file located at file_path
+    """
+    try:
+        with open(data.path, 'r') as f:
+            return f.read()
+    except Exception as e:
+        raise Exception(f"Error reading {data.path}\n{e}")
+
+
+def put_page_content(content:str, data_uri: str, **kwargs) -> str:
+    """
+    Write content to a file at relative_path (relative to data_path).
+    relative_path is a list of strings, each string is a directory name 
+    representing a directory tree.
+    handle s3 or file storage, use '/' as separator for s3 key and os.sep
+    as separtors for file storage
+    """
+    data = parse_data_uri(data_uri)
+
+    if data.store == 's3':
+        return put_s3_content(data, content, **kwargs)
+    elif data.store == 'file':
+        return put_file_content(data, content)
+    else:
+        raise Exception(f"Unknown data store: {data.store}")
+
+
+def put_s3_content(data: DataStorage, content, **kwargs) -> str:
+    """
+    Write content to an object named data.path
+    """
+    s3 = boto3.client('s3', **kwargs)
+    s3.put_object(
+        ACL='bucket-owner-full-control',
+        Bucket=data.bucket,
+        Key=data.path,
+        Body=content
+    )
+    return data.uri
+
+
+def put_file_content(data: DataStorage, content) -> str:
+    """
+    Write content to a file at data.path
+    """
+    file_path = os.sep.join(data.path.split('/'))
+    directory_path = os.path.dirname(file_path)
+    if not os.path.exists(directory_path):
+        os.makedirs(directory_path)
```

Comment:
This needs to be:

`os.makedirs(directory_path, exist_ok=True)`

to avoid getting a FileExistsError when one of the intermediate dirs already exists, which happens when writing the 2nd page of data.

Reply:
Strange, I don't get that error. Will do.